### PR TITLE
[WIP] Feat: Adding DynamoDB BatchWriteItem throttling property as expected

### LINKED ISSUE #3879: Dynamodb BatchWriteItem not Throttling Properly
# Type of request: This is a ...

[ X ] bug report
[ ] feature request

# Detailed description

On BatchWriteItem if items are throttled, we are expected to receive a list of `UnprocessedItems` in the response instead of HTTP 500

## Expected behavior

Receive HTTP 200 but with Unprocessed Items in the response: https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchWriteItem.html#DDB-BatchWriteItem-response-UnprocessedItems

## Actual behavior

Receive HTTP 500 ProvisionedThroughPutException. "All or nothing"

# Steps to reproduce

Start LocalStack with DynamoDb Service and "DYNAMODB_ERROR_PROBABILITY=1.0"

## Command used to start LocalStack

```
new LocalStackContainer(DockerImageName.parse("localstack/localstack:0.12.9"))
      .withEnv("DYNAMODB_ERROR_PROBABILITY", "1.0")
      .withServices(DYNAMODB)
```

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)

You can just Randomly generate a big list of objects (exceeding 1kb per batch) and check this behaviour

### PR DESCRIPTION
**Pull request by mgagliardo**
Fixes #3879.

* Pretty nice example to make this "fail" if you add `DYNAMODB_WRITE_ERROR_PROBABILITY=1.0` and use [this guide](
https://www.dynamodbguide.com/working-with-multiple-items/) to see it fail.
* Some minor fixes on config, i.e. `DYNAMODB_WRITE_ERROR_PROBABILITY` and `DYNAMODB_READ_ERROR_PROBABILITY` where being "captured" on the `config` file but even if you passed it it would not work.

* I have also found line `70` (function `action_should_throttle`) on `dynamodb_listener` not working as it would expect the action (i.e. `BatchWriteItem`) to match `DynamoDB_20120810.BatchWriteItem` when being run from the AWS command line, as the CLI would just pass the action, not the action_prefix.
Fixed with:
```
        return (action in throttled) or (action in actions)
```

As usual comments are welcome.

### CODE REVIEW

File: localstack/services/dynamodb/dynamodb_listener.py [hunk 1]
```diff
@@ -319,11 +327,21 @@ def return_response(self, method, path, data, headers, response):
...
-            records = self.prepare_batch_write_item_records(record, data)
+            records, unprocessed_items = self.prepare_batch_write_item_records(record, data)
             for record in records:
                 event_sources_or_streams_enabled = (event_sources_or_streams_enabled or
                     has_event_sources_or_streams_enabled(record['eventSourceARN'], streams_enabled_cache))
+            if response.status_code == 200 and len(unprocessed_items) > 0:
+                content = json.loads(to_str(response.content))
+                table_name = list(data['RequestItems'].keys())[0]
+                if table_name not in content['UnprocessedItems']:
+                    content['UnprocessedItems'][table_name] = []
```
Comment (driverpt) [inline localstack/services/dynamodb/dynamodb_listener.py:340]: The proper response is:

```json
"UnprocessedItems": {
  "<TableName>": {
     "DeleteRequest":{
        "Key1": {<Key1>},
        "Key2": {<Key2>}
     },
     "PutRequest": {
        "Item1": { Item1 },
        "Item2": { Item2 }
     }
  }
}
```

File: localstack/services/dynamodb/dynamodb_listener.py [hunk 2]
```diff
@@ -152,13 +153,19 @@ def forward_request(self, method, path, data, headers):
...
                         inner_request = request.get(key)
                         if inner_request:
-                            existing_items.append(find_existing_item(inner_request, table_name))
+                            if self.should_throttle(action):
+                                unprocessed_items.append(inner_request)
+                            else:
+                                item = find_existing_item(inner_request, table_name)
+                                existing_items.append(item)
             ProxyListenerDynamoDB.thread_local.existing_items = existing_items
+            ProxyListenerDynamoDB.thread_local.unprocessed_items = unprocessed_items
```
Comment (driverpt) [inline localstack/services/dynamodb/dynamodb_listener.py:168]: Suggestion:

Split `unprocessed_items` into `unprocessed_put_items` and `unprocessed_delete_items` to make it easier to distribute amongst the response `dict`

File: tests/integration/test_error_injection.py [hunk 3]
```diff
@@ -99,6 +99,13 @@ def test_dynamodb_write_error_injection():
     config.DYNAMODB_WRITE_ERROR_PROBABILITY = 1.0
     assert_raises(ClientError, table.put_item, Item={PARTITION_KEY: short_uid(), 'data': 'foobar123'})
 
+    # with a probability of 1, BatchWriteItem does not throw errors but logs UnprocessedItems
```
Comment (driverpt) [inline tests/integration/test_error_injection.py:102]: `BatchWriteItem` throws `ProvisionedThroughputExceededException` if ALL items in Batch are Throttled

File: localstack/services/dynamodb/dynamodb_listener.py (hunk 1 above)
Comment (mgagliardo) [reply localstack/services/dynamodb/dynamodb_listener.py]: Sorry but acording to https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchWriteItem.html#DDB-BatchWriteItem-response-UnprocessedItems the right syntax would be

```
   "UnprocessedItems": { 
      "string" : [ 
         { 
            "DeleteRequest": { 
               "Key": { 
```

So a list, not a JSON object

File: localstack/services/dynamodb/dynamodb_listener.py [hunk 4]
```diff
@@ -319,11 +332,23 @@ def return_response(self, method, path, data, headers, response):
...
                     has_event_sources_or_streams_enabled(record['eventSourceARN'], streams_enabled_cache))
+            if response.status_code == 200 and any(unprocessed_items):
+                content = json.loads(to_str(response.content))
+                table_name = list(data['RequestItems'].keys())[0]
+                if table_name not in content['UnprocessedItems']:
+                    content['UnprocessedItems'][table_name] = []
+                for key in ['PutRequest', 'DeleteRequest']:
+                    if any(unprocessed_items[key]):
+                        content['UnprocessedItems'][table_name].append({key: unprocessed_items[key]})
+                print(content)
```
Comment (whummer) [inline localstack/services/dynamodb/dynamodb_listener.py:349]: nitpick: can we please remove the `print(..)` before merging

Comment (mgagliardo) [reply localstack/services/dynamodb/dynamodb_listener.py]: FML, my bad Waldemar. Fixed

**Comment by whummer** [comment]
Thanks! 👍

**Comment by driverpt** [comment]
Tyvm! Good job

**Pull request merged**